### PR TITLE
:sparkles: Apple Silicon (arm64) support, fixes #36

### LINKED ISSUE #36: Architecture errors on macosx-10.9-universal2-3.10/lru.o
Hi when installing `web3` on a 2017 mac installation breaks with multiple `Unsupported architecture` errors.

Fix:
`ARCHFLAGS="-arch x86_64"  pip install web3`

Full stack:
```
      building 'lru' extension
      creating build
      creating build/temp.macosx-10.9-universal2-3.10
      clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -arch arm64 -arch x86_64 -g -I/Users/jeanpaulruizdepraz/Documents/software/sni/web3_py/venv/include -I/Library/Frameworks/Python.framework/Versions/3.10/include/python3.10 -c lru.c -o build/temp.macosx-10.9-universal2-3.10/lru.o
      In file included from lru.c:1:
      In file included from /Library/Frameworks/Python.framework/Versions/3.10/include/python3.10/Python.h:11:
      In file included from /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/include/limits.h:21:
      In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/limits.h:63:
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/cdefs.h:807:2: error: Unsupported architecture
      #error Unsupported architecture
       ^
      In file included from lru.c:1:
      In file included from /Library/Frameworks/Python.framework/Versions/3.10/include/python3.10/Python.h:11:
      In file included from /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/include/limits.h:21:
      In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/limits.h:64:
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/machine/limits.h:8:2: error: architecture not supported
      #error architecture not supported
       ^
      In file included from lru.c:1:
      In file included from /Library/Frameworks/Python.framework/Versions/3.10/include/python3.10/Python.h:25:
      In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/stdio.h:64:
      In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_stdio.h:71:
      In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_types.h:27:
      In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h:33:
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/machine/_types.h:34:2: error: architecture not supported
      #error architecture not supported
       ^
      In file included from lru.c:1:
      In file included from /Library/Frameworks/Python.framework/Versions/3.10/include/python3.10/Python.h:25:
      In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/stdio.h:64:
      In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_stdio.h:71:
      In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_types.h:27:
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h:55:9: error: unknown type name '__int64_t'
      typedef __int64_t       __darwin_blkcnt_t;      /* total blocks */
              ^
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h:56:9: error: unknown type name '__int32_t'; did you mean '__int128_t'?
      typedef __int32_t       __darwin_blksize_t;     /* preferred block size */
              ^
      note: '__int128_t' declared here
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h:57:9: error: unknown type name '__int32_t'; did you mean '__int128_t'?
      typedef __int32_t       __darwin_dev_t;         /* dev_t */
              ^
      note: '__int128_t' declared here
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h:60:9: error: unknown type name '__uint32_t'; did you mean '__uint128_t'?
      typedef __uint32_t      __darwin_gid_t;         /* [???] process and group IDs */
              ^
      note: '__uint128_t' declared here
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h:61:9: error: unknown type name '__uint32_t'; did you mean '__uint128_t'?
      typedef __uint32_t      __darwin_id_t;          /* [XSI] pid_t, uid_t, or gid_t*/
              ^
      note: '__uint128_t' declared here
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h:62:9: error: unknown type name '__uint64_t'
      typedef __uint64_t      __darwin_ino64_t;       /* [???] Used for 64 bit inodes */
              ^
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h:68:9: error: unknown type name '__darwin_natural_t'
      typedef __darwin_natural_t __darwin_mach_port_name_t; /* Used by mach */
              ^
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h:70:9: error: unknown type name '__uint16_t'; did you mean '__uint128_t'?
      typedef __uint16_t      __darwin_mode_t;        /* [???] Some file attributes */
              ^
      note: '__uint128_t' declared here
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h:71:9: error: unknown type name '__int64_t'
      typedef __int64_t       __darwin_off_t;         /* [???] Used for file sizes */
              ^
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h:72:9: error: unknown type name '__int32_t'; did you mean '__int128_t'?
      typedef __int32_t       __darwin_pid_t;         /* [???] process and group IDs */
              ^
      note: '__int128_t' declared here
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h:73:9: error: unknown type name '__uint32_t'; did you mean '__uint128_t'?
      typedef __uint32_t      __darwin_sigset_t;      /* [???] signal set */
              ^
      note: '__uint128_t' declared here
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h:74:9: error: unknown type name '__int32_t'; did you mean '__int128_t'?
      typedef __int32_t       __darwin_suseconds_t;   /* [???] microseconds */
              ^
      note: '__int128_t' declared here
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h:75:9: error: unknown type name '__uint32_t'; did you mean '__uint128_t'?
      typedef __uint32_t      __darwin_uid_t;         /* [???] user IDs */
              ^
      note: '__uint128_t' declared here
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h:76:9: error: unknown type name '__uint32_t'; did you mean '__uint128_t'?
      typedef __uint32_t      __darwin_useconds_t;    /* [???] microseconds */
              ^
      note: '__uint128_t' declared here
      In file included from lru.c:1:
      In file included from /Library/Frameworks/Python.framework/Versions/3.10/include/python3.10/Python.h:25:
      In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/stdio.h:64:
      In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_stdio.h:71:
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_types.h:43:9: error: unknown type name '__uint32_t'; did you mean '__uint128_t'?
      typedef __uint32_t      __darwin_wctype_t;
              ^
      note: '__uint128_t' declared here
      In file included from lru.c:1:
      In file included from /Library/Frameworks/Python.framework/Versions/3.10/include/python3.10/Python.h:25:
      In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/stdio.h:64:
      In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_stdio.h:75:
      In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_va_list.h:31:
      /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/machine/types.h:37:2: error: architecture not supported
      #error architecture not supported
       ^
      fatal error: too many errors emitted, stopping now [-ferror-limit=]
      20 errors generated.
      error: command '/usr/bin/clang' failed with exit code 1
      [end of output]
```

### PR DESCRIPTION
**Pull request by AndreMiras**
Note that we (cross) compile for both arm64 and the universal2. This may feel redundant, it can still be interesting to have both. While universal2 is more convenient arm64 is lighter.

Also drop Python 3.7 (end-of-life 2023-06-27).
Add Python 3.12 and pypy-3.10 support.

Bump CI actions versions as well.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -1,16 +1,19 @@
 name: Tests
 
-on: [push]
+on: [push, pull_request]
```
Comment (AndreMiras) [inline .github/workflows/tests.yml:3]: Also sneaked that change in so we would have caught that issue earlier:
https://github.com/amitdev/lru-dict/pull/49#issuecomment-1666815977

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
+
+env:
+  PYTHON_VERSIONS: ["3.8", "3.9", "3.10", "3.11", "3.12", "pypy-3.8", "pypy-3.9", "pypy-3.10"]
```
Comment (AndreMiras) [inline .github/workflows/tests.yml:6]: DRY fix, but also removed 3.7, and added 3.12 as well as pypy-3.10.
But if we prefer I can address these changes in another dedicated PR

File: .github/workflows/build-and-deploy.yml [hunk 3]
```diff
@@ -8,21 +8,26 @@ jobs:
...
-        uses: pypa/cibuildwheel@v2.12.0
+        uses: pypa/cibuildwheel@v2.16.2
         env:
           # configure cibuildwheel to build native archs ('auto'), and some emulated ones
           CIBW_ARCHS_LINUX: auto aarch64 ppc64le
+          CIBW_ARCHS_MACOS: x86_64 arm64 universal2
+          CIBW_TEST_REQUIRES: .[test]
+          CIBW_TEST_COMMAND: pytest {project}
+          # arm64 cannot be tested on a x86_64 CI runner
+          CIBW_TEST_SKIP: "*-macosx_arm64"
```
Comment (AndreMiras) [inline .github/workflows/build-and-deploy.yml:30]: This is the macOS arm64 support part, all the rest of the PR is just chore

**Comment by AndreMiras** [comment]
Hi @amitdev, how can I help you merge this one?
Do you want me to split it up into smaller concerns or describe further what's doing and why it's needed?
I'm excited to see it merged as it would make life easier to Apple ARM/M1 developers

**Comment by amitdev** [comment]
Thanks @AndreMiras for the PR. I was away for sometime so couldn't get this merged. Will get a release out soon.

**Comment by amitdev** [comment]
@AndreMiras can you fix this [issue](https://github.com/amitdev/lru-dict/actions/runs/6526334419) with tests btw? If not I'll take a look when I get time later:
```
[Invalid workflow file: .github/workflows/tests.yml#L6](https://github.com/amitdev/lru-dict/actions/runs/6526334419/workflow)
The workflow is not valid. .github/workflows/tests.yml (Line: 6, Col: 20): A sequence was not expected .github/workflows/tests.yml (Line: 13, Col: 17): Unrecognized named-value: 'env'. Located at position 10 within expression: fromJson(env.PYTHON_VERSIONS)
```

**Comment by AndreMiras** [comment]
Here's the fix https://github.com/amitdev/lru-dict/pull/55, thanks for the heads up.

**Pull request merged**